### ISSUE TITLE
Generic linux support

### ISSUE THREAD

**Issue by t3nsed** [oxen-io/lokinet#909]
The [documentation](https://docs.loki.network/Lokinet/Guides/lokinet-linux-guide/) currently only supports instructions for Ubuntu and Ubuntu-like linux distros. Building on any other distro takes time since there is no documentation on dependencies and building tools used, thus boils down to `make` spam to understand what is missing.

I'm aware that supporting multiple linux distros can easily become a hassle for projects of this size, but a list of dependencies and generic build instructions to get a binary would be very helpful.

**Comment by majestrate** [comment]
generic linux needs:

* libsodium >= 1.0.17
* libuv >= 1.27.0
* cmake >= 3.6.0
* GNU make
* setcap

to build:

    $ mkdir build && cd build && cmake .. && make

to install:

    # make install

this should be in the readme if it isn't already.

**Comment by t3nsed** [comment]
> this should be in the readme if it isn't already.

I've used the readme, and it sort of is:

```
Build requirements:

GNU Make
CMake
C++ 17 capable C++ compiler
gcovr (if generating test coverage with gcc)
libuv >= 1.27.0
```

`libsodium` and `setcap` are not mentioned explicitly in the readme, but can be inferred by reading the `make` errors. Not sure if you would want to add those.

My issue was that these were not mentioned in the [docs at loki.network](https://docs.loki.network/Lokinet/Guides/lokinet-linux-guide/). Its more of a usability thing, some users might not care about the development on github thus only look and follow the instructions in the docs and not those in this repo.

**Comment by majestrate** [comment]
libsodium isn't required for the current master branch but once `0.6.0` is merged into master it will.

**Comment by zangetsuu66** [comment]
hi i cant compile on fedora 32 

CMake Warning (dev) at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:272 (message):
  The package name passed to `find_package_handle_standard_args` (sodium)
  does not match the name of the calling package (Sodium).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  cmake/FindSodium.cmake:227 (find_package_handle_standard_args)
  crypto/CMakeLists.txt:71 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

**Comment by majestrate** [comment]
what version of libsodium is installed?

**Comment by zangetsuu66** [comment]
libsodium 1.0.18 rev 3.fc.32 include in fedora 32

**Comment by majestrate** [comment]
if i recall the build also needs `pkg-config`

**Comment by unmellow** [comment]
after downloading the git repository  
making the build directory running git submodule update --init --recursive
in that directory then running cmake .. -DBUILD_STATIC_DEPS=ON -DBUILD_SHARED_LIBS=OFF -DSTATIC_LINK=ON
successfully i get this error when runing make
```
[  2%] Built target openssl_external
[  4%] Built target zlib_external
[  6%] Built target expat_external
[  8%] Built target unbound_external
[ 10%] Built target sodium_external
[ 12%] Built target libuv_external
[ 15%] Built target zmq_external
[ 15%] Performing install step for 'curl_external'
  CC       vquic/libcurl_la-ngtcp2.lo
vquic/ngtcp2.c:27:10: fatal error: ngtcp2/ngtcp2_crypto.h: No such file or directory
   27 | #include <ngtcp2/ngtcp2_crypto.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:2991: vquic/libcurl_la-ngtcp2.lo] Error 1
make[2]: *** [CMakeFiles/curl_external.dir/build.make:107: static-deps-sources/src/curl_external-stamp/curl_external-install] Error 2
make[1]: *** [CMakeFiles/Makefile2:694: CMakeFiles/curl_external.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```
i tried installing ngtcp2-notls-git (plain ngtcp2 fails to install)
i still get this error

i also had to install other unmentioned dendency's like nghttp3

running make -j$(nproc)
just compiles other submodules successfully at the same time as ngtcp2 fails
the same error with different things around it

**Comment by majestrate** [comment]
this is with the aur package i assume?

**Comment by unmellow** [comment]
no i tried the aur package and it had it's own problems and i didn't know if it was official
so i thought i would try doing things the "right way" first (compiling from source following directions as closely as possible)
but i am using manjaro

the other issues have nothing to do with compiling also I'm not really sure what's going on at all
but i know it deserves it's own issue
(as in i managed to compile but the lokinet daemon always timed out during startup)
(and running lokinet -g && lokinet-bootstrap then lokinet always resulted in the same error after i stopped the daemon of course)

**Comment by majestrate** [comment]
> no i tried the aur package and it had it's own problems and i didn't know if it was official

it's not official.

> so i thought i would try doing things the "right way" first (compiling from source following directions as closely as possible)
> but i am using manjaro
> 
> the other issues have nothing to do with compiling also I'm not really sure what's going on at all
> but i know it deserves it's own issue

please make additional issues if you can

> (as in i managed to compile but the lokinet daemon always timed out during startup)

timed out how? if you are running with systemd install libsystemd and recompile for sd_notify support.

> (and running lokinet -g && lokinet-bootstrap then lokinet always resulted in the same error after i stopped the daemon of course)

which result specifically?

**Comment by unmellow** [comment]
ok made a separate issue here #1630
for the lokinet executable problem

**Comment by Saladin1812** [comment]
It is not official but it uses official tar.gz from this github page releases as a source you can verify the **source files** and the **PKGBUILD** here and I request @majestrate to do so (verify source files as listed as "Sources" in these AUR packages) -

https://aur.archlinux.org/packages/lokinet-bin/   -  for the binary version
 https://aur.archlinux.org/packages/lokinet/   - this one compiles
 
In my case there were no isses I am on Arch Linux. @unmellow  You can also comment on the AUR to ask the package maintainer.

**Comment by majestrate** [comment]
I personally don't maintain any packaging of any kind, I sign tags and tarballs.

@jagerman maintains the debian packages and a community member has unofficial RPM package. 

There are some community members that want to support arch linux so they do. If you want packages for your package manager then feel free to package it but I doubt i'll be of any use in that process. I don't know how arch linux packages work I haven't used arch in many years so my review is not useful.

**Comment by majestrate** [comment]
@shuriken1812 / @unmellow we have fixes for AUR packages https://github.com/oxen-io/lokinet/tree/makepkg/contrib/archlinux

**Comment by PixsaOJ** [comment]
@majestrate `lokinet` package on AUR is missing PGP keys. I cannot `gpg --key-recive ID` from any key server I know.

**Comment by majestrate** [comment]
see https://github.com/oxen-io/lokinet/discussions/1823
